server/src/main/java/org/opensearch/index/codec/composite/Composite99DocValuesReader.java

```diff
@@ -29,11 +60,123 @@
  */
 @ExperimentalApi
 public class Composite99DocValuesReader extends DocValuesProducer implements CompositeIndexReader {
-    private DocValuesProducer delegate;
+    private static final Logger logger = LogManager.getLogger(CompositeIndexMetadata.class);
+
+    private final DocValuesProducer delegate;
+    private IndexInput dataIn;
+    private ChecksumIndexInput metaIn;
+    private final Map<String, Tree> starTreeMap = new LinkedHashMap<>();
+    private final Map<String, CompositeIndexMetadata> compositeIndexMetadataMap = new LinkedHashMap<>();
+    private final Map<String, DocValuesProducer> compositeDocValuesProducerMap = new LinkedHashMap<>();
+    private final List<CompositeIndexFieldInfo> compositeFieldInfos = new ArrayList<>();
+    private final SegmentReadState readState;
 
-    public Composite99DocValuesReader(DocValuesProducer producer, SegmentReadState state) throws IOException {
+    public Composite99DocValuesReader(DocValuesProducer producer, SegmentReadState readState) throws IOException {
         this.delegate = producer;
-        // TODO : read star tree files
+        this.readState = readState;
+
+        String metaFileName = IndexFileNames.segmentFileName(
+            readState.segmentInfo.name,
+            readState.segmentSuffix,
+            Composite99DocValuesFormat.META_EXTENSION
+        );
+
+        String dataFileName = IndexFileNames.segmentFileName(
+            readState.segmentInfo.name,
+            readState.segmentSuffix,
+            Composite99DocValuesFormat.DATA_EXTENSION
+        );
+
+        boolean success = false;
+        try {
+
+            // initialize meta input
+            dataIn = readState.directory.openInput(dataFileName, readState.context);
+            CodecUtil.checkIndexHeader(
+                dataIn,
+                Composite99DocValuesFormat.DATA_CODEC_NAME,
+                Composite99DocValuesFormat.VERSION_START,
+                Composite99DocValuesFormat.VERSION_CURRENT,
+                readState.segmentInfo.getId(),
+                readState.segmentSuffix
+            );
+
+            // initialize data input
+            metaIn = readState.directory.openChecksumInput(metaFileName, readState.context);
+            Throwable priorE = null;
+            try {
+                CodecUtil.checkIndexHeader(
+                    metaIn,
+                    Composite99DocValuesFormat.META_CODEC_NAME,
+                    Composite99DocValuesFormat.VERSION_START,
+                    Composite99DocValuesFormat.VERSION_CURRENT,
+                    readState.segmentInfo.getId(),
+                    readState.segmentSuffix
+                );
+
+                while (true) {
```

Comment:
This is dangerous, can we instead use the entire file's length or similar data as an indicator ?

Reply:
It's used similarly in Lucene90PointsReader. Any exception should be handled via try / catch. I will check if any other way is possible